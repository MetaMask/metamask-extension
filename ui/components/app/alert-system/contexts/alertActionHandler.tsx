import React, {
  ReactElement,
  createContext,
  useCallback,
  useContext,
} from 'react';

type AlertActionHandlerContextType = {
  processAction: (actionKey: string) => void;
};

export const AlertActionHandlerContext = createContext<
  AlertActionHandlerContextType | undefined
>(undefined);

export const AlertActionHandlerProvider: React.FC<{
  children: ReactElement;
  onProcessAction: (actionKey: string) => void;
}> = ({ children, onProcessAction }) => {
  const processAction = useCallback(
    (_actionKey: string) => {
      onProcessAction(_actionKey);
    },
    [onProcessAction],
  );

  return (
    <AlertActionHandlerContext.Provider value={{ processAction }}>
      {children}
    </AlertActionHandlerContext.Provider>
  );
};

export const useAlertActionHandler = () => {
  const context = useContext(AlertActionHandlerContext);
  if (!context) {
    throw new Error(
      'useAlertActionHandler must be used within an AlertActionHandlerProvider',
    );
  }
  return context;
};
