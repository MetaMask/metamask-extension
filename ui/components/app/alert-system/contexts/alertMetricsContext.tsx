import React, { createContext, useContext, ReactNode, useMemo } from 'react';

type TrackFunctionType = (alertKey: string) => void;

const AlertMetricsContext = createContext<{
  trackAlertActionClicked: TrackFunctionType;
  trackAlertRender: TrackFunctionType;
  trackInlineAlertClicked: TrackFunctionType;
} | null>(null);

type AlertMetricsProps = {
  children: ReactNode;
  metrics: {
    trackAlertActionClicked: TrackFunctionType;
    trackAlertRender: TrackFunctionType;
    trackInlineAlertClicked: TrackFunctionType;
  };
};

export const AlertMetricsProvider: React.FC<AlertMetricsProps> = ({
  children,
  metrics,
}) => {
  const { trackAlertActionClicked, trackAlertRender, trackInlineAlertClicked } =
    metrics;

  const value = useMemo(
    () => ({
      trackAlertActionClicked,
      trackAlertRender,
      trackInlineAlertClicked,
    }),
    [trackAlertActionClicked, trackAlertRender, trackInlineAlertClicked],
  );

  return (
    <AlertMetricsContext.Provider value={value}>
      {children}
    </AlertMetricsContext.Provider>
  );
};

export const useAlertMetrics = () => {
  const context = useContext(AlertMetricsContext);
  if (!context) {
    throw new Error(
      'useAlertMetrics must be used within an AlertMetricsProvider',
    );
  }
  return context;
};
