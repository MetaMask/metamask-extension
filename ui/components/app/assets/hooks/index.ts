export { default as useAssetListTokenDetection } from './useAssetListTokenDetection';
export { default as useNativeTokenBalance } from './useNativeTokenBalance';
export { default as useNetworkFilter } from './useNetworkFilter';
export { default as usePrimaryCurrencyProperties } from './usePrimaryCurrencyProperties';
export { default as useTokenDisplayInfo } from './useTokenDisplayInfo';
