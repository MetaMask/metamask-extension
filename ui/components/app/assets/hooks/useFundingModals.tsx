import { useState } from 'react';

const useFundingModals = () => {
  const [showFundingMethodModal, setShowFundingMethodModal] = useState(false);
  const [showReceiveModal, setShowReceiveModal] = useState(false);

  const onClickReceive = () => {
    setShowFundingMethodModal(false);
    setShowReceiveModal(true);
  };
};

export default useFundingModals;
