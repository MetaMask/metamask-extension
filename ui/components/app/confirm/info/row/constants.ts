export const TEST_ADDRESS = '0x5CfE73b6021E818B776b421B1c4Db2474086a7e1';

export enum RowAlertKey {
  EstimatedFee = 'estimatedFee',
  SigningInWith = 'signingInWith',
  RequestFrom = 'requestFrom',
  Resimulation = 'resimulation',
  Speed = 'speed',
}

export enum AlertActionKey {
  Buy = 'buy',
  ShowAdvancedGasFeeModal = 'showAdvancedGasModal',
  ShowGasFeeModal = 'showGasFeeModal',
}
