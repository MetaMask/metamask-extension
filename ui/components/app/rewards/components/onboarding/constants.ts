export const REWARDS_ONBOARD_OPTIN_LEGAL_LEARN_MORE_URL =
  'https://support.metamask.io/manage-crypto/metamask-rewards';

export const REWARDS_ONBOARD_TERMS_URL = 'https://go.metamask.io/rewards-terms';
