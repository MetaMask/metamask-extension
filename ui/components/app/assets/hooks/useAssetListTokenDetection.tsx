import { useContext, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NetworkConfiguration } from '@metamask/network-controller';
import { Token } from '../types';
import {
  getAllDetectedTokensForSelectedAddress,
  getDetectedTokensInCurrentNetwork,
  getIsTokenNetworkFilterEqualCurrentNetwork,
  getSelectedAddress,
  getUseTokenDetection,
} from '../../../../selectors';
import { importAllDetectedTokens } from '../util/importAllDetectedTokens';
import {
  getCurrentChainId,
  getNetworkConfigurationsByChainId,
  getSelectedNetworkClientId,
} from '../../../../../shared/modules/selectors/networks';
import { MetaMetricsContext } from '../../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
  MetaMetricsTokenEventSource,
} from '../../../../../shared/constants/metametrics';
import {
  AssetType,
  TokenStandard,
} from '../../../../../shared/constants/transaction';
import { addImportedTokens } from '../../../../store/actions';

const useAssetListTokenDetection = () => {
  const trackEvent = useContext(MetaMetricsContext);
  const dispatch = useDispatch();
  const detectedTokensMultichain: {
    [key: string]: Token[];
  } = useSelector(getAllDetectedTokensForSelectedAddress);
  const networkClientId = useSelector(getSelectedNetworkClientId);
  const selectedAddress = useSelector(getSelectedAddress);
  const useTokenDetection = useSelector(getUseTokenDetection);
  const currentChainId = useSelector(getCurrentChainId);
  const isOnCurrentNetwork = useSelector(
    getIsTokenNetworkFilterEqualCurrentNetwork,
  );
  const detectedTokens = useSelector(getDetectedTokensInCurrentNetwork) || [];
  const allNetworks: Record<`0x${string}`, NetworkConfiguration> = useSelector(
    getNetworkConfigurationsByChainId,
  );

  const multichainDetectedTokensLength = Object.values(
    detectedTokensMultichain || {},
  ).reduce((acc, tokens) => acc + tokens.length, 0);

  const [showDetectedTokens, setShowDetectedTokens] = useState(false);

  const handleAddImportedTokens = async (
    tokens: Token[],
    networkClientIdProp: string,
  ) => {
    await dispatch(addImportedTokens(tokens as Token[], networkClientIdProp));
  };

  const trackTokenAddedEvent = (importedToken: Token, chainId: string) => {
    trackEvent({
      event: MetaMetricsEventName.TokenAdded,
      category: MetaMetricsEventCategory.Wallet,
      sensitiveProperties: {
        token_symbol: importedToken.symbol,
        token_contract_address: importedToken.address,
        token_decimal_precision: importedToken.decimals,
        source: MetaMetricsTokenEventSource.Detected,
        token_standard: TokenStandard.ERC20,
        asset_type: AssetType.token,
        token_added_type: 'detected',
        chain_id: chainId,
      },
    });
  };

  // Add detected tokens to sate
  useEffect(() => {
    if (!useTokenDetection) {
      return;
    }

    importAllDetectedTokens(
      isOnCurrentNetwork,
      detectedTokensMultichain,
      allNetworks,
      networkClientId,
      currentChainId,
      detectedTokens,
      handleAddImportedTokens,
      trackTokenAddedEvent,
    );
  }, [
    isOnCurrentNetwork,
    selectedAddress,
    networkClientId,
    detectedTokens.length,
    multichainDetectedTokensLength,
  ]);

  return { showDetectedTokens, setShowDetectedTokens };
};

export default useAssetListTokenDetection;
