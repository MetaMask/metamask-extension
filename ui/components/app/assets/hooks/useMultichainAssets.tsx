import { useSelector } from 'react-redux';
import { Hex } from '@metamask/utils';
import { getMultichainSelectedAccountCachedBalance } from '../../../../selectors/multichain';
import { getSelectedInternalAccount } from '../../../../selectors';
import {
  TranslateFunction,
  networkTitleOverrides,
} from '../util/networkTitleOverrides';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import { formatWithThreshold } from '../util/formatWithThreshold';
import { getIntlLocale } from '../../../../ducks/locale/locale';
import { getCurrentCurrency } from '../../../../ducks/metamask/metamask';
import {
  MULTICHAIN_PROVIDER_CONFIGS,
  MultichainNetworks,
} from '../../../../../shared/constants/multichain/networks';
import { TokenWithFiatAmount } from '../types';
import { getMultiChainAssets } from '../../../../selectors/assets';

const useMultiChainAssets = () => {
  const t = useI18nContext();
  const locale = useSelector(getIntlLocale);
  const selectedAccount = useSelector(getSelectedInternalAccount);
  const currentCurrency = useSelector(getCurrentCurrency);

  const multichainAssets = useSelector((state) =>
    getMultiChainAssets(state, selectedAccount),
  );

  // the following condition is needed to satisfy e2e check-balance.spec.ts
  // this is because the new multichain data is not being mocked within the withSolanaAccountSnap test fixture
  // balances render as expected without this condition during local testing
  const cachedBalance = useSelector(getMultichainSelectedAccountCachedBalance);
  if (cachedBalance === 0) {
    return [
      {
        chainId: MultichainNetworks.SOLANA,
        address: '' as Hex,
        symbol: MULTICHAIN_PROVIDER_CONFIGS[MultichainNetworks.SOLANA].ticker,
        string: `${cachedBalance} ${currentCurrency}`,
        primary: cachedBalance,
        image: '',
        secondary: cachedBalance,
        tokenFiatAmount: cachedBalance,
        isNative: true,
        decimals: 9, // hard coded decimal value
        title: MULTICHAIN_PROVIDER_CONFIGS[MultichainNetworks.SOLANA].nickname,
        isStakeable: false,
      },
    ];
  }

  return multichainAssets.map((asset: TokenWithFiatAmount) => {
    const fiatAmount = formatWithThreshold(asset.secondary, 0.01, locale, {
      style: 'currency',
      currency: currentCurrency.toUpperCase(),
    });

    return {
      ...asset,
      title: asset.isNative
        ? networkTitleOverrides(t as TranslateFunction, {
            title: asset.title,
          })
        : asset.title,

      secondary: fiatAmount, // secondary balance (usually in fiat)
    };
  });
};

export default useMultiChainAssets;
