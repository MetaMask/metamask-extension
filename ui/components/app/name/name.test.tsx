import * as React from 'react';
import { NameType } from '@metamask/name-controller';
import configureStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import { MetaMetricsContext } from '../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
} from '../../../../shared/constants/metametrics';
import { useDisplayName } from '../../../hooks/useDisplayName';
import { mockNetworkState } from '../../../../test/stub/networks';
import { CHAIN_IDS } from '../../../../shared/constants/network';
import Name from './name';

jest.mock('../../../hooks/useDisplayName');

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
}));

const ADDRESS_NO_SAVED_NAME_MOCK = '0xc0ffee254729296a45a3885639ac7e10f9d54977';
const ADDRESS_SAVED_NAME_MOCK = '0xc0ffee254729296a45a3885639ac7e10f9d54979';
const SAVED_NAME_MOCK = 'TestName';
const VARIATION_MOCK = 'testVariation';

const STATE_MOCK = {
  metamask: {
    ...mockNetworkState({ chainId: CHAIN_IDS.MAINNET }),
  },
};

describe('Name', () => {
  const store = configureStore()(STATE_MOCK);
  const useDisplayNameMock = jest.mocked(useDisplayName);

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders when no address value is passed', () => {
    useDisplayNameMock.mockReturnValue({
      name: null,
      hasPetname: false,
    });

    const { container } = renderWithProvider(
      <Name
        type={NameType.ETHEREUM_ADDRESS}
        value={''}
        variation={VARIATION_MOCK}
      />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders address with no saved name', () => {
    useDisplayNameMock.mockReturnValue({
      name: null,
      hasPetname: false,
    });

    const { container } = renderWithProvider(
      <Name
        type={NameType.ETHEREUM_ADDRESS}
        value={ADDRESS_NO_SAVED_NAME_MOCK}
        variation={VARIATION_MOCK}
      />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders address with saved name', () => {
    useDisplayNameMock.mockReturnValue({
      name: SAVED_NAME_MOCK,
      hasPetname: true,
    });

    const { container } = renderWithProvider(
      <Name
        type={NameType.ETHEREUM_ADDRESS}
        value={ADDRESS_SAVED_NAME_MOCK}
        variation={VARIATION_MOCK}
      />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders address with image', () => {
    useDisplayNameMock.mockReturnValue({
      name: SAVED_NAME_MOCK,
      hasPetname: true,
      image: 'test-image',
    });

    const { container } = renderWithProvider(
      <Name
        type={NameType.ETHEREUM_ADDRESS}
        value={ADDRESS_SAVED_NAME_MOCK}
        variation={VARIATION_MOCK}
      />,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  describe('metrics', () => {
    // @ts-expect-error This is missing from the Mocha type definitions
    it.each([
      ['saved', ADDRESS_SAVED_NAME_MOCK, true],
      ['not saved', ADDRESS_NO_SAVED_NAME_MOCK, false],
    ])(
      'sends displayed event with %s name',
      async (_: string, value: string, hasPetname: boolean) => {
        const trackEventMock = jest.fn();

        useDisplayNameMock.mockReturnValue({
          name: hasPetname ? SAVED_NAME_MOCK : null,
          hasPetname,
        });

        renderWithProvider(
          <MetaMetricsContext.Provider value={trackEventMock}>
            <Name
              type={NameType.ETHEREUM_ADDRESS}
              value={value}
              variation={VARIATION_MOCK}
            />
          </MetaMetricsContext.Provider>,
          store,
        );

        expect(trackEventMock).toHaveBeenCalledWith({
          event: MetaMetricsEventName.PetnameDisplayed,
          category: MetaMetricsEventCategory.Petnames,
          properties: {
            petname_category: NameType.ETHEREUM_ADDRESS,
            has_petname: hasPetname,
          },
        });
      },
    );
  });
});
