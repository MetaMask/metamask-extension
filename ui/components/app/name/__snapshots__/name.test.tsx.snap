// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name renders address with image 1`] = `
<div>
  <div
    class="mm-box mm-box--display-flex"
  >
    <div
      class="name name__saved"
    >
      <img
        alt=""
        class="identicon"
        src="test-image"
        style="height: 16px; width: 16px; border-radius: 8px;"
      />
      <p
        class="mm-box mm-text name__name mm-text--body-md mm-box--color-text-default"
      >
        TestName
      </p>
    </div>
  </div>
</div>
`;

exports[`Name renders address with no saved name 1`] = `
<div>
  <div
    class="mm-box mm-box--display-flex"
  >
    <div
      class="name name__missing"
    >
      <span
        class="mm-box name__icon mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/question.svg');"
      />
      <p
        class="mm-box mm-text name__value mm-text--body-md mm-box--color-text-default"
      >
        0xC0ffe...54977
      </p>
    </div>
  </div>
</div>
`;

exports[`Name renders address with saved name 1`] = `
<div>
  <div
    class="mm-box mm-box--display-flex"
  >
    <div
      class="name name__saved"
    >
      <div
        class=""
      >
        <div
          class="identicon"
          style="height: 16px; width: 16px; border-radius: 8px;"
        >
          <div
            style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 16px; height: 16px; display: inline-block; background: rgb(245, 196, 0);"
          >
            <svg
              height="16"
              width="16"
              x="0"
              y="0"
            >
              <rect
                fill="#C8144D"
                height="16"
                transform="translate(-0.8205726313476182 -2.375300755278998) rotate(408.5 8 8)"
                width="16"
                x="0"
                y="0"
              />
              <rect
                fill="#FB1860"
                height="16"
                transform="translate(7.733089412585649 -1.3465466417019656) rotate(402.6 8 8)"
                width="16"
                x="0"
                y="0"
              />
              <rect
                fill="#03505E"
                height="16"
                transform="translate(1.1000464230208766 -13.147318085284851) rotate(308.0 8 8)"
                width="16"
                x="0"
                y="0"
              />
            </svg>
          </div>
        </div>
      </div>
      <p
        class="mm-box mm-text name__name mm-text--body-md mm-box--color-text-default"
      >
        TestName
      </p>
    </div>
  </div>
</div>
`;

exports[`Name renders when no address value is passed 1`] = `
<div>
  <div
    class="mm-box mm-box--display-flex"
  >
    <div
      class="name name__missing"
    >
      <span
        class="mm-box name__icon mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/question.svg');"
      />
      <p
        class="mm-box mm-text name__value mm-text--body-md mm-box--color-text-default"
      />
    </div>
  </div>
</div>
`;
